### PR TITLE
PayTrace: Adjust purchase and capture methods to handle MultiResponse…

### PR DESCRIPTION
**Pull request by meagabeth**
… scenarios

- Adjust refund method to include money argument
- Revise message_from method to include error message details and handle multiple errors
- Revise authorization_from to support returning a third_party_token for store
- Revise necessary tests

Local:
4787 tests, 73769 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
Unit:
19 tests, 74 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
Remote:
27 tests, 68 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

### CODE REVIEW

File: lib/active_merchant/billing/gateways/pay_trace.rb [hunk 1]
```diff
@@ -43,23 +43,19 @@ def initialize(options = {})
...
 
       def purchase(money, payment_or_customer_id, options = {})
-        post = {}
-        add_amount(post, money, options)
-        if customer_id?(payment_or_customer_id)
-          post[:customer_id] = payment_or_customer_id
-          endpoint = ENDPOINTS[:customer_id_sale]
+        if visa_or_mastercard?(options)
+          MultiResponse.run(:use_first_response) do |r|
+            customer_id?(payment_or_customer_id) ? endpoint = [ENDPOINTS][:customer_id_sale] : endpoint = ENDPOINTS[:keyed_sale]
```
Comment (therufs) [inline lib/active_merchant/billing/gateways/pay_trace.rb:48]: We can (I think) tidy this up by putting `endpoint =` at the beginning like so: 

```
endpoint = customer_id?(payment_or_customer_id) ? [ENDPOINTS][:customer_id_sale] : ENDPOINTS[:keyed_sale]
```

Comment (meagabeth) [reply lib/active_merchant/billing/gateways/pay_trace.rb]: Revised! Thanks for teaching me a thing with that!

File: lib/active_merchant/billing/gateways/pay_trace.rb [hunk 2]
```diff
@@ -165,15 +162,35 @@ def acquire_access_token
...
+        else
+          add_payment(post, payment_or_customer_id)
+          add_address(post, payment_or_customer_id, options)
+          add_customer_data(post, options)
+        end
+
+        post
+      end
+
+      def capture_request(money, authorization, options)
```
Comment (therufs) [inline lib/active_merchant/billing/gateways/pay_trace.rb:179]: this is a totally nitpicky thing to notice or mention but can we have either both or neither of these start with `build_`?

File: test/remote/gateways/remote_pay_trace_test.rb [hunk 3]
```diff
@@ -16,9 +16,10 @@ class RemotePayTraceTest < Test::Unit::TestCase
   def setup
     @gateway = PayTraceGateway.new(fixtures(:pay_trace))
 
-    @amount = 100
+    @amount = 1000
     @credit_card = credit_card('4012000098765439')
     @mastercard = credit_card('5499740000000057')
+    @invalid_card = credit_card('54545454545454', month: '14', year: '1999')
```
Comment (therufs) [inline test/remote/gateways/remote_pay_trace_test.rb:22]: TONIGHT WE'RE GONNA PARTY LIKE IT'S

File: test/remote/gateways/remote_pay_trace_test.rb [hunk 4]
```diff
@@ -269,15 +277,15 @@ def test_refund_without_amount
...
     assert_success refund
     assert_equal 'Your transaction was successfully refunded.', refund.message
   end
 
   def test_failed_refund
-    response = @gateway.refund('', @options)
+    response = @gateway.refund(100, '', @options)
     assert_failure response
-    assert_equal 'One or more errors has occurred.', response.message
+    assert_equal 'Errors- code:981, message:["Log in failed for insufficient permissions."]', response.message
```
Comment (therufs) [inline test/remote/gateways/remote_pay_trace_test.rb:288]: :squint: this seems like an unexpected error message in this context?

File: test/remote/gateways/remote_pay_trace_test.rb (hunk 3 above)
Comment (meagabeth) [reply test/remote/gateways/remote_pay_trace_test.rb:22]: exactly

File: test/remote/gateways/remote_pay_trace_test.rb (hunk 4 above)
Comment (meagabeth) [reply test/remote/gateways/remote_pay_trace_test.rb:288]: Per our chat, this has been revised!

File: lib/active_merchant/billing/gateways/pay_trace.rb (hunk 2 above)
Comment (meagabeth) [reply lib/active_merchant/billing/gateways/pay_trace.rb:179]: Love it - yes and done!

File: lib/active_merchant/billing/gateways/pay_trace.rb [hunk 5]
```diff
@@ -79,21 +75,22 @@ def authorize(money, payment_or_customer_id, options = {})
...
-        return response unless visa_or_mastercard?(options)
-
-        send_level_3_data(response, options)
+        if visa_or_mastercard?(options)
+          MultiResponse.run do |r|
+            r.process { commit(ENDPOINTS[:capture], build_capture_request(money, authorization, options)) }
+            r.process { commit(ENDPOINTS[:"level_3_#{options[:visa_or_mastercard]}"], send_level_3_data(r, options)) }
+          end
+        else
+          post = build_build_capture_request(money, authorization, options)
```
Comment (therufs) [inline lib/active_merchant/billing/gateways/pay_trace.rb:84]: oh_no.gif

Comment (meagabeth) [reply lib/active_merchant/billing/gateways/pay_trace.rb]: I DO A FIX 🔨

**Review by therufs — APPROVED** [review]
tests pass && lgtm!! 🎉 🚢

File: lib/active_merchant/billing/gateways/pay_trace.rb (hunk 5 above)
Comment (meagabeth) [reply lib/active_merchant/billing/gateways/pay_trace.rb]: vs code replace feature did me a big ole' fail

**Pull request merged**